### ISSUE TITLE
Recommended Express integration

### ISSUE THREAD

**Issue by namoscato** [taskforcesh/bullmq#982]
We have a relatively straightforward integration of this library on Heroku (that generally follows [this guide](https://devcenter.heroku.com/articles/node-redis-workers)):

1. `worker` process – small Node.js process that instantiates a `Worker` instance, leveraging [throng](https://github.com/hunterloftis/throng) / clustering
1. `web` process – Express server instantiates a singleton ioredis connection and `Queue` instance; in practice, some of our Express routes add a job to the queue

We're instantiating the ioredis connection like so:

```ts
new IORedis(redisUrl, {
  maxRetriesPerRequest: null,
  enableReadyCheck: false,
});
```

per the recommendation added in https://github.com/taskforcesh/bullmq/pull/836.

**Is this the recommended Express setup?** We're specifically concerned that disabling `maxRetriesPerRequest` will pose issues in our web context where ideally the number of retries would be finite (per [ioredis' default](https://github.com/luin/ioredis#auto-reconnect)) and commands would "fail fast" to deliver a timely HTTP response.

**Comment by manast** [comment]
This is a very interesting point and I am afraid I may have overlooked it. Trying to add to a queue that has lost a connection currently does not fail, which in most use cases is not what you want. So it seems like the optimal settings for a worker are not the same as to a client adding jobs to the queue... I will try to find a solution for this.

**Comment by manast** [comment]
Still I am not sure what would be the proper way to solve this. So we want the call to add to fail fast, but we also want the queue connection to reconnect to the Redis server as soon as that server is online/reachable again, so we need infinite retries but also failing the command immediately if not currently connected, do you know which ioredis settings would accomplish this?

**Comment by manast** [comment]
Ok, so after digging a bit more into it, it is possible we could just enforce the null on maxRetriesPerRequest only on connections that use "blocking" commands, that way we will fail fast when calling BullMQs api and still keep the blocking commands running stable during disconnection episodes-

**Comment by namoscato** [comment]
Thanks for the quick reply, @manast!

> we need infinite retries but also failing the command immediately if not currently connected, do you know which ioredis settings would accomplish this?

Maybe this is what you are getting at, but I don't think you can accomplish this with a single ioredis instance / connection. Certainly their [default `maxRetriesPerRequest: 20`](https://github.com/luin/ioredis/blob/v4.28.2/lib/redis/RedisOptions.ts#L84) would accomplish the latter.

Just to clarify, why does the `Queue` class need a persistent Redis connection? Or perhaps said differently, why can't we solely rely on ioredis' finite reconnection algorithm that would presumably happen when adding a job to the queue?

I'm happy to do some more research on how other frameworks handle this when I have time. I also admittedly do not have a clear mental model yet on how this library manages the Redis connection.

**Comment by manast** [comment]
I think it is important to separate between getting an error when calling a queue operation on a disconnected client vs having infinite retries for connection, because actually, in your express application you do not want to need to restart the whole application when the 20 retries have been expired and all the calls to the queue will now fail even if the Redis service is up and running again. So I do not think the current settings are wrong, I need to investigate this a bit further before I know how to proceed.

**Comment by manast** [comment]
Could you try in your Queue connection settings to disable ioredis offline queue to see if it behaves the way you want? ```enableOfflineQueue``` from https://github.com/luin/ioredis/blob/master/API.md

**Comment by divine** [comment]
Hello,

I'm having the same issue after #836, Queue is just stuck and doesn't add any queues at all.

> Could you try in your Queue connection settings to disable ioredis offline queue to see if it behaves the way you want? `enableOfflineQueue` from https://github.com/luin/ioredis/blob/master/API.md

Doesn't help...
```
[CRIT] [-] [core] Error: Stream isn't writeable and enableOfflineQueue options is false
[CRIT] [-] [core]     at Redis.sendCommand (/node_modules/ioredis/built/redis/index.js:659:24)
[CRIT] [-] [core]     at Redis.subscribe (/node_modules/ioredis/built/commander.js:122:25)
[CRIT] [-] [core]     at new Lock (/node_modules/ioredfour/lib/ioredfour.js:50:31)
```

Thanks!

**Comment by manast** [comment]
@divine I think you should post with the specifics of your use case, and if you can provide a piece of code that reproduces the issue the chances are much much higher that we can help you. 

The ```enableOfflineQueue``` is for a different use case.

**Comment by divine** [comment]
> @divine I think you should create a new issue with the specifics of your use case, and if you can provide a piece of code that reproduces the issue the chances are much much higher that we can help you.
> 
> The `enableOfflineQueue` is for a different use case.

I'll try reproducing it with minimal code and create a new issue ASAP, just want to clarify:

> doesn't add any queues at all.

It adds after some time until it reconnects back which makes the whole queue add lagging for a few seconds sometimes minutes...

Thank you for your reply.

**Comment by manast** [comment]
> doesn't add any queues at all.

Okok, so you mean it does not add any jobs at all. I think I misunderstood you.
When you say it is lagging, what happens during that lag period when you add jobs, does it give an error or it just keeps waiting for the connection to be ready?

**Comment by divine** [comment]
> > doesn't add any queues at all.
> 
> Okok, so you mean it does not add any jobs at all. I think I misunderstood you.

Sorry about that!

> When you say it is lagging, what happens during that lag period when you add jobs, does it give an error or it just keeps waiting for the connection to be ready?

It doesn't give any error it just keeps waiting for the connection to be ready. Bullmq Queue is used as a Haraka plugin on a email server for notifications which makes it feel "lag"* instead of continuing processing thousand of emails per minute...

The important logic for me is that there should be probably a constant check if the queue is ready, queue.add should just process it later if the connection isn't ready or just fail instead of waiting for reconnection. Any help is appreciated!

Also sometimes there are might be packet drops in the private network connection to Redis for a few seconds (thanks God for cloud providers internal network which is going crazy sometimes for no reason).

\* Doesn't accept new connections as the current email transaction isn't processed fully.

Thanks!

**Comment by manast** [comment]
> The important logic for me is that there should be probably a constant check if the queue is ready, queue.add should just process it later if the connection isn't ready or just fail instead of waiting for reconnection. Any help is appreciated!

Normally the robust way to implement this kind of functionality is to have a retry mechanism, so when "add fails" you keep retrying for some time until it succeeds. When this happens of course the system will be slower, on the other hand, we should expect this as a very exceptional scenario, normally the servers are up and running.

Anyway, I will investigate why add is not failing when the connection is not available as it should when passing the ```enableOfflineQueue``` setting.

**Comment by manast** [comment]
Ok, so I have tested this scenario and I think it works in an acceptable way. Basically, the limitation is that if Redis is not online when you instantiate the queue AND you add a job before it has become available it will not fail directly, however, if Redis has been available and then it goes offline when it is back again and you try to add a job it will fail as expected. Here is a piece of test code that can be used for verifying this:
```
const { Queue } = require("bullmq");

const name = process.argv.slice(2);
const videoQueue = new Queue("video transcoding", {
  connection: {
    enableOfflineQueue: false,
  },
});

function delay(time) {
  return new Promise((resolve) => {
    setTimeout(resolve, time);
  });
}

async function add() {
  try {
    console.log("waiting");
    await delay(10000);
    console.log("Done waiting");
    await videoQueue.add("wall", { name });
    console.log("Job added");
  } catch (err) {
    console.log("We failed", err);
  }
}

add();
```

If you run this code with an available Redis instance and wait a couple of seconds, and then you shut down the instance (before the 10 seconds delay has passed) you will see that it fails as it should when trying to add the job.

This is probably as good as it gets for now, but in the future we may improve the startup code so that it works more consistenly.

**Comment by manast** [comment]
I updated the documentation with a pattern for this case: https://docs.bullmq.io/patterns/failing-fast-when-redis-is-down

I think this should resolve the issue.

**Comment by namoscato** [comment]
@manast, thanks for researching this!

I tested `enableOfflineQueue=false`, and it's certainly better than retrying indefinitely.

I do think a _small_ number of retries (via `maxRetriesPerRequest`) would be more robust, handling network blips and whatnot. Would you ever consider supporting such configuration within the context of the `Queue` class?

**Comment by manast** [comment]
> I do think a _small_ number of retries (via `maxRetriesPerRequest`) would be more robust, handling network blips and whatnot. Would you ever consider supporting such a configuration within the context of the `Queue` class?

Yes. We could support that setting for the connections that are used for the "interactive" functionality of the queue.

**Comment by manast** [comment]
Oh wait. I may have answered a bit too quickly. The "maxRetriesPerRequest" as an ioredis setting will not improve the situation as a user of the Queue class. The problem is that when the retries are exhausted, then that queue instance will not work anymore, since the connection is broken forever. So what you are asking is a "retry" mechanism at the command level, i.e. when calling for example "add", if the connection is momentarily broken, we retry the "add" a few times until actually failing. Implementing such a feature in BullMQ would imply a retry mechanism for all the calls we do to Redis, so it seems like the wrong place to put this functionality, instead, this should be implemented at ioredis level, which is kind of what "enableOfflineQueue" does... my suggestion if you want a more robust add is that you implement such a retry mechanism on top of your "add" call instead, same way as you would implement a retry mechanism when doing REST calls to an external API.

**Comment by namoscato** [comment]
> So what you are asking is a "retry" mechanism at the command level

Yeah, exactly.

From what I can tell, this is the purpose of `maxRetriesPerRequest`. Per [their documentation](https://github.com/luin/ioredis#auto-reconnect):

> By default, all pending commands will be flushed with an error every 20 retry attempts. That makes sure commands won't wait forever when the connection is down. You can change this behavior by setting `maxRetriesPerRequest`

I tested this behavior with:

```ts
import express from "express";
import IORedis from "ioredis";

const redis = new IORedis(process.env.REDIS_URL, {
  maxRetriesPerRequest: 5,
});

const app = express();

app.post("/redis", async (req, res) => {
  try {
    const pong = await redis.ping();
    res.status(200).json(pong);
  } catch (err) {
    res.status(503).send(err);
  }
});
```

1. Start all services
2. POST `/redis` returns "PONG"
3. Stop Redis
4. POST `/redis` returns a 503 after 5 retries (~1s)

    ```
    MaxRetriesPerRequestError: Reached the max retries per request limit (which is 20). Refer to "maxRetriesPerRequest" option for details.
    ```

    Note that ioredis continues to log connect-related errors (presumably trying to reconnect?):

    ```
    Error: connect ECONNREFUSED 127.0.0.1:6379
        at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1159:16) {
      errno: -61,
      code: 'ECONNREFUSED',
      syscall: 'connect',
      address: '127.0.0.1',
      port: 6379
    }
    ```

5. Restart Redis
6. POST `/redis` returns "PONG" again, and ioredis connect error logs stop

When `maxRetriesPerRequest: null`, the step (4) request hangs indefinitely.

**Comment by namoscato** [comment]
@manast, just wanted to check back in to see if you have given any more thought into supporting a finite `maxRetriesPerRequest`?

**Comment by manast** [comment]
@namoscato if it is as simple as just to enable a non null ```maxRetriesPerRequest``` then I do not have anything against in supporting it if enables your use-case. It would be great iff you could provide a PR for this though :)

**Comment by hardcodet** [comment]
FYI, at this point, bull still displays a deprecation warning during startup, which caused me to apply that setting and probably stick to it unless we run into issues (that may be very hard to track down).

```
BullMQ: DEPRECATION WARNING! Your redis options maxRetriesPerRequest must be null and enableReadyCheck false. On the next versions having this settings will throw an exception
```

With regards to this thread: Will the recommendation still stand to have `mayRetriesPerRequest` to be `null`?
In any case, may I suggest that your recommendation makes it into the docs? And more importantly: If the answer is "it depends", it would be great if the docs could educate us mere redis mortals ;)

**Comment by manast** [comment]
It is actually documented but in the new documentation site we are preparing: https://docs.bullmq.io/bull/patterns/persistent-connections
Let me know if this responds to your question.

**Comment by hardcodet** [comment]
Very nicely written, thank you @manast !

In case that's not planned anyway, might be worth linking to from https://docs.bullmq.io/guide/connections?

**Comment by hardcodet** [comment]
Addendum: So I am setting up to different connections with different settings for queue and worker. Which leaves the question which one of the two is now more suitable for the queue scheduler. I guess we only want to queue to fail fast, but maybe document that one, too?

**Comment by namoscato** [comment]
https://github.com/taskforcesh/bullmq/pull/1119 takes a pass at removing these problematic overrides for the `Queue` class. I imagine this does not need to affect other classes, but @hardcodet poses a fair question.

I read through https://github.com/OptimalBits/bull/issues/1873 (where all of this started) and am still pretty confused about all of this.

**Comment by manast** [comment]
> Addendum: So I am setting up to different connections with different settings for queue and worker. Which leaves the question which one of the two is now more suitable for the queue scheduler. I guess we only want to queue to fail fast, but maybe document that one, too?

The QueueScheduler requires a permanent collection, so it will need the same settings as the Worker class. 

I am preparing documentation and fixes for all these inconsistencies.

**Comment by manast** [comment]
> I read through [OptimalBits/bull#1873](https://github.com/OptimalBits/bull/issues/1873) (where all of this started) and am still pretty confused about all of this.

These are old issues they probably will confuse you more than give you any clarification. It is quite simple, for connections that need are persistent and should never fail you use the ioredis setting "maxRetriesPerRequest: null", this is normally the case of a Worker, that is consuming jobs "forever" and should never fail. But for the Queue class, for example if you are "adding" a job, you would like to get an exception if the call did not work after a certain number of retries, default for ioredis is 20.
